proposals/dgl-operator-proposal.md

```diff
@@ -0,0 +1,384 @@
+    1. Single-Pod, single-thread partitioning, _will call SSP or `DGL-API` later_
+    2. Single-Pod, multi-process partitioning (native ParMETIS version), _will call SMP or `ParMETIS` later_
+    3. Multi-Pod, single-thread partitioning (may need some adoptions in ParMETIS), _will call MSP or `DistParMETIS` later_
+2. Working Pod definitions:
+    1. Launcher: A very light weight Pod traces the whole DGL workflow on the Pod level.
+    2. Partitioner: A Pod does the graph partitioning.
+    3. Worker: One of the Pods does the DGL training.
+3. `dglrun`: A DGL workflow script file, used to take the control of graph partitioning, distributed training and some utils on the container level.
+4. `watcher-loop`: A variant of [`kubectl-delivery`](https://github.com/kubeflow/mpi-operator/tree/master/cmd/kubectl-delivery), which can track not only the readiness of all workloads and also the completion of all workloads. Moreover, `kubectl` part is removed.
+5. `kubectl-download`: A tiny image that only does `kubectl` part of [`kubectl-delivery`](https://github.com/kubeflow/mpi-operator/tree/master/cmd/kubectl-delivery), downloading the `kubectl` binary file.
+6. `ipconfig`: A text file lists the worker pods' name, ip, port (used in training phase of dgl server) and the available slots (GPUs) in each pod.
+7. `kubexec.sh`: A helper shell script that can invoke `kubectl exec` for remote execution
+
+## Goals
+
+A Kubeflow user should be able to run training using DGL as easily as then can using Tensorflow or PyTorch. This proposal is centered around a Kubernetes operator for DGL. A user should be able to run single worker job with DGL (1 launcher, 1 partitioner and 1 worker), distributed training jobs with DGL and `DGL-API` partitioning mode (1 launcher, 1 partitioner and multiple workers), distributed training jobs with DGL and `ParMETIS` partitioning mode (1 launcher, 1 partitioner and multiple workers), and distributed training jobs with DGL and `DistParMETIS` partitioning mode (1 launcher and multiple workers).
+
+This proposal defines the following:
+
+- A DGL operator
+- A way to deploy the DGL Operator
+- A single worker DGL example
+- A distributed DGL example with `DGL-API` partitioning mode
+- A distributed DGL example with `ParMETIS` partitioning mode
+- A distributed DGL example with `DistParMETIS` partitioning mode
+
+## Non-Goals
+
+Currently, for the scope of this proposal, we won't be addressing the method for serving the model.
+
+## API
+
+### Container Image
+
+- A requirement on user's container image is that it includes DGL, related libraries, and our workflow scripts (`dglrun` and utils).
+- SSH is not needed (or used).
+### Custom Resource Definition
+
+The CRD yaml example is as following:
+
+```yaml
+apiVersion: qihoo.net/v1alpha1
+kind: DGLJob
+metadata:
+  name: dgl-graphsage
+  namespace: dgl-operator
+spec:
+  cleanPodPolicy: Running
```

Comment:
Hi Xiaoyu and comminuty : )  I don't know if it is probably better to use runPolicy field https://github.com/kubeflow/common/blob/master/pkg/apis/common/v1/types.go#L176 as other operators did instead of directly using cleanPodPolicy ?

For example tfjobs https://github.com/kubeflow/training-operator/blob/master/pkg/apis/tensorflow/v1/types.go#L55; 
pytorchjob https://github.com/kubeflow/training-operator/blob/master/pkg/apis/pytorch/v1/types.go#L53. 

For standalone operator, this works fine; but for all-in-one training-operator, all of controllers use the same base job controller, where the clean up logic try to read runPolicy.cleanPodPolicy.  Like mpi-operator migrating into training-operator, the fields mismatched will be an headache.